Subject: allow less-than-3-component kernel version number
From: Michael Tokarev <mjt@tls.msk.ru>
Forwarded: https://bugs.busybox.net/show_bug.cgi?id=5444
Bug: http://bugs.debian.org/684611

Busybox expects at least 3-component linux version number,
and segfaults if less than 3 components are available.

--- debian.orig/libbb/kernel_version.c
+++ debian/libbb/kernel_version.c
@@ -21,18 +21,19 @@
 {
 	struct utsname name;
 	char *s;
-	int i, r;
+	int r;
 
 	if (uname(&name) == -1) {
 		bb_perror_msg("can't get system information");
 		return 0;
 	}
 
-	s = name.release;
-	r = 0;
-	for (i = 0; i < 3; i++) {
-		r = r * 256 + atoi(strtok(s, "."));
-		s = NULL;
+	r = atoi(strtok(name.release, ".")) * 256 * 256;
+	if ((s = strtok(NULL, ".")) != NULL) {
+		r |= atoi(s) * 256;
+		if ((s = strtok(NULL, ".")) != NULL)
+			r |= atoi(s);
 	}
+
 	return r;
 }
